src/lib/util/src/mem.rs

```diff
@@ -0,0 +1,102 @@
+use core::slice;
+
+use log::{print, println};
+
+use crate::mmio::{read32, write32};
+
+// NOTE: Print an empty line _before_ instead of doing println!(), such that
+// after progress dots, there is always a line break before the error message.
+fn check32(addr: usize, val: u32) -> Result<(), ()> {
+    let v = read32(addr);
+    if v != val {
+        print!("\nError @ {addr:08x}: expected {val:08x}, got {v:08x}");
+        Err(())
+    } else {
+        Ok(())
+    }
+}
+
+const TEST_PATTERN0: u32 = 0x2233_ccee;
+const TEST_PATTERN1: u32 = 0x5577_aadd;
+const TEST_PATTERN2: u32 = 0x1144_bbff;
+const TEST_PATTERN3: u32 = 0x6688_9900;
+
+pub fn test(base: usize, size: usize) {
+    let mut error_count = 0;
+    let limit = base + size;
+    let step_size = 0x10;
+    // print 64 steps, which gets slower with a higher size to test
+    let print_step = size / step_size / 64;
+
+    println!("Memory test: {size}bytes @ 0x{base:016x}");
+
+    println!("Memory test: write patterns...");
+    for (i, a) in (base..limit).step_by(step_size).enumerate() {
+        if i % print_step == 0 {
+            print!(".");
+        }
+        #[allow(clippy::identity_op)]
```

Comment:
I have not tried this, but could you do
let i = I as u32
and then just use I in the next four lines?

Reply:
e.g. 

fn main() {
    let i: u32 = 0;
    println!("Hello, world!{i}");
    let i = i as i32;
    println!("Hello, world!{i}");
}

Reply:
We could... an extra line/assignment vs another inline cast - I'm sure the compiler would optimize it either way.
I personally find the few little inlines okay here.